features/device_key/source/DeviceKey.cpp

```diff
@@ -0,0 +1,287 @@
+    return DEVICEKEY_SUCCESS;
+}
+
+int DeviceKey::get_derived_key(uint32_t *ikey_buff, size_t ikey_size, const unsigned char *isalt,
+                               size_t isalt_size, unsigned char *output, uint32_t ikey_type)
+{
+    //KDF in counter mode implementation as described in Section 5.1
+    //of NIST SP 800-108, Recommendation for Key Derivation Using Pseudorandom Functions
+    int ret;
+    size_t counter = 0;
+    char separator = 0x00;
+    mbedtls_cipher_context_t ctx;
+    unsigned char output_len_enc[ 4 ] = {0};
+    unsigned char counter_enc[ 1 ] = {0};
+
+    DEVKEY_WRITE_UINT32_LE(output_len_enc, ikey_type);
+
+    mbedtls_cipher_type_t mbedtls_cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
+    if (DEVICE_KEY_32BYTE == ikey_size) {
+        mbedtls_cipher_type = MBEDTLS_CIPHER_AES_256_ECB;
+    }
+
+    const mbedtls_cipher_info_t *cipher_info = mbedtls_cipher_info_from_type(mbedtls_cipher_type);
+
+    mbedtls_cipher_init(&ctx);
+    ret = mbedtls_cipher_setup(&ctx, cipher_info);
+    if (ret != 0) {
+        goto finish;
+    }
+
+    do {
+
+        ret = mbedtls_cipher_cmac_starts(&ctx, (unsigned char *)ikey_buff, ikey_size * 8);
+        if (ret != 0) {
+            goto finish;
+        }
+
+        DEVKEY_WRITE_UINT8_LE(counter_enc, (counter+1));
+
+        ret = mbedtls_cipher_cmac_update(&ctx, (unsigned char *)counter_enc, sizeof(counter_enc));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_update(&ctx, isalt, isalt_size);
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_update(&ctx, (unsigned char *)&separator, sizeof(char));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_update(&ctx, (unsigned char *)&output_len_enc, sizeof(output_len_enc));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        ret = mbedtls_cipher_cmac_finish(&ctx, output + (DEVICE_KEY_16BYTE * (counter)));
+        if (ret != 0) {
+            goto finish;
+        }
+
+        counter++;
+
+    } while (DEVICE_KEY_16BYTE * counter < ikey_type);
+
+finish:
+    mbedtls_cipher_free( &ctx );
+
+    if (DEVICEKEY_SUCCESS != ret) {
+        return DEVICEKEY_ERR_CMAC_GENERIC_FAILURE;
+    }
+
+    return DEVICEKEY_SUCCESS;
+}
+
+int DeviceKey::generate_key_by_trng(uint32_t *output, size_t& size)
+{
+#if defined(DEVICE_TRNG)
+    size_t in_size;
+    size_t ongoing_size;
+    size_t final_size;
+    trng_t trng_obj;
+    int ret = DEVICEKEY_SUCCESS;
+    unsigned char *pBuffer = (unsigned char *)output;
+
+    memset(output, 0, size);
+
+    if (DEVICE_KEY_16BYTE > size) {
+        return DEVICEKEY_BUFFER_TOO_SMALL;
+    } else if (DEVICE_KEY_16BYTE <= size && DEVICE_KEY_32BYTE > size) {
+        size = DEVICE_KEY_16BYTE;
+    } else {
+        size = DEVICE_KEY_32BYTE;
+    }
+
+    trng_init(&trng_obj);
+
+    final_size = 0;
+    in_size = size;
+    while (final_size < size) {
```

Comment:
This can be simplified to `in_size > 0`, removing `final_size` altogether. The modification of `size` is unnecessary in the end, because the function guarantees the requested amount of random data to be provided in case of success. We can therefore change the signature of this function to `generate_key_by_trng( unsigned char *output, size_t size )` -- passing a pointer/reference suggests that the function might generate less data than requested (e.g. `trng_get_bytes`), but that's not the case here.

Reply:
ok